dags/utils/slack_operator.py

```diff
@@ -131,6 +113,123 @@ def extract_exception_from_log(log_text):
     return None, None
 
 
+def extract_all_exceptions_from_log(log_text):
+    """
+    Extract all Python exceptions from raw Airflow logs. 
+    Returns list of tuples (exception_type, exception_message, source).
+    Source is "ETL" for exceptions before "Task failed with exception" line,
+    "Airflow" for exceptions after that line.
+    """
+    
+    exceptions = []
+    lines = log_text.split('\n')
+    
+    # Find the "Task failed with exception" dividing line
+    task_failed_line_idx = None
+    for i, line in enumerate(lines):
+        if "Task failed with exception" in line:
```

Comment:
This particular string is the boundary between:

* Our code: exceptions that are coming out of our containerized, inside-the-black-box python apps that do our ETL work. 
* Not our code: exceptions that are coming out of Airflow itself, such as its code that starts and manages docker containers.